### PR TITLE
Refactors lavaland to be minecraft

### PR DESCRIPTION
**Pull request by Kmc2000**
## About The Pull Request

I present to you: 
TG:MC

A revolutionary new gaming experience which will totally revolutionise the way YOU play miner

## Why It's Good For The Game

This PR refactors a pretty unpopular feature, replacing it with something everyone can get behind!
Suffer not a miner to fall in lava, unless they dig straight down...of course!

This also allows for massive creativity, just like the source material, as you can place any block you mine :)

Thank you for your attention

FOOTAGE:
https://i.gyazo.com/c84655cbfd757b6850feec67f3fa6c76.mp4

Demo of construction mechanics:
https://streamable.com/82yle

## Changelog
:cl: Kmc2000
refactor: Lavaland is now minecraft
/:cl:

### CODE REVIEW

**Comment by Manly-Dwarf** [comment]
Highest quality /tg/station PRs. Proof that not all /tg/coders are grudgecoding frauds.

**Comment by HomocidlePancakes** [comment]
PR this again before april fools along with baymed port

**Comment by Teepeepee** [comment]
make putting them back down have a delay too, otherwise they will be used to cover your escape in maint with a wall that takes some time to get down which also requires you to have a pick

**Comment by Moja10** [comment]
Yeah no brickspamming and this could be an actual PR

**Comment by Nickvr628** [comment]
Every day we get closer to Fortnite13

**Comment by datorangebottle** [comment]
The most epic PR of the century.

Screw Ripley reworks, screw economy, screw the tesla engine and removal of security borgs. This PR is the one that changes the face of /tg/ forever.

**Comment by SouDescolado** [comment]
> make putting them back down have a delay too, otherwise they will be used to cover your escape in maint with a wall that takes some time to get down which also requires you to have a pick

This is how it works in minecraft tho

**Comment by Kmc2000** [comment]
One day.
I will surpass radial menus.

Until that day
I can only dream of such perfection

**Comment by SouDescolado** [comment]
Btw, you need to fix that when you destroy dirt it converts the ground under it to grass.

**Comment by Kmc2000** [comment]
Astute obvservation gamer!

**Comment by HomocidlePancakes** [comment]
Are diamond swords and armor included? Also replace drakes with ender dragons while you are at it

**Comment by Kmc2000** [comment]
If oranges doesn't immediately close this I'll consider it.

**Comment by SouDescolado** [comment]
No, I mean it cannot convert the station to grass. But yea, it should be dirt under the dirt block.

**Comment by 2dplanner** [comment]
My name is 2dplanner and this is my favourite pr on the citadel.

**Comment by Kmc2000** [comment]
That's what I meant, Sou

**Comment by Shadowflame909** [comment]
Wait I just realized. Lavaland is a black variant of the asteroid. But with lava.

**Comment by Kmc2000** [comment]
All the more reason to replace it with minecraft

**Comment by SouDescolado** [comment]
@Kmc2000 Welp, I guess you know what you are doing. If we are going full minecraft, make stone become cobblestone when broken, cobblestone not generated on map and if you put cobblestone on a furnace thing(?) it becomes stone. I think we have an ore processor, but I dont play miner very often to remember.

This way you can dig your way and make your own untraceable home, otherwise it is clear that someone tempered with the walls

**Comment by Kmc2000** [comment]
I mean
I can make this as realistic as people want it.

What's to stop us from making the nether and end, too?

**Comment by SouDescolado** [comment]
You made it so if you break the wall it can generate ores. There is nothing stopping you to break a block (Any strong block, like andesite for example), grab ores, then place it again and break it for more ores. Either make it not works on placed blocks or make it destroy the block when ores get generated(which will still allow you to farm blocks then take your time placing and breaking)

**Comment by Repukan** [comment]
Nice I hate it

**Comment by Teepeepee** [comment]
make it so the necropolis entrance becomes the nether portal when you kill legion

**Comment by Rohesie** [comment]
We can make this work. +1

**Comment by Moja10** [comment]
While you are at it KMC can you add sec borgs back in too?

**Review by ExcessiveUseOfCobblestone — DISMISSED** [review]
has just the right amount of cobblestone

(meme review)

**Comment by Kmc2000** [comment]
We're in business boys 😎

**Comment by Moja10** [comment]
craftbench component when

**Comment by 81Denton** [comment]
I don't even know what the FUCK is going on anymore

**Review by 81Denton — DISMISSED** [review]

**Comment by Daxxed** [comment]
Looks fun to tide with.
Selling pickaxes pm me

**Comment by 2dplanner** [comment]
make skeletons and zombies respawn in uniluminated areas

**Comment by Keekenox** [comment]
My god this is amazing. KMC I fucking love you.

**Comment by wesoda25** [comment]
Who woulda thunk the sound of destroying dirt/mining stone would be so nostalgic

**Comment by nicbn** [comment]
Good
Now replace the craft menu with the Minecraft workbench thing

**Comment by IndieanaJones** [comment]
Could we actually add this as another z level, where miners can choose to go to either lavaland or minecraft world?

**Comment by SouDescolado** [comment]
I think if Goof was still alive he would say something like "Muh Trademarks", so maybe name it like Craftland or Fortcraft or something generic like that (And avoid a group of 13 year olds googling minecraft online server and finding us)

**Comment by Kmc2000** [comment]
Holy shit craftland 13 that's genius

**Comment by optimumtact** [comment]
you could make this a VR level i suppose.

High quality shitpost though.

**Comment by ExcessiveUseOfCobblestone** [comment]
> I think if Goof was still alive he would say something like "Muh Trademarks", so maybe name it like Craftland or Fortcraft or something generic like that (And avoid a group of 13 year olds googling minecraft online server and finding us)

if you care about that stuff then the sprites (prob sound, I didn't look at it) need to be botched too.

**Comment by adamkad1** [comment]
this is pretty good

**Comment by Whoneedspacee** [comment]
VR Hub is coming, make it an away mission map and I’ll package it into the map that spawns every round 👍

**Comment by Kmc2000** [comment]
How do I do that @whoneedspacee?
I'll make a proper minecraft map in that case, with nether and end :b1:

**Comment by Whoneedspacee** [comment]
Just make any away mission map by copying the current vr maps with the spawns and everythin and makin your map based off of that. Then add the file path to the away missions config. Might wanna look at my megafauna vr pr or another vr pr to get an idea of what you need to do.

Then I’ll just shove it into my main map when I pr my thing since I’m pretty much having all the small things on the same map with portals to them. I wouldn’t make it too big otherwise it’ll be too big of a map and remain on rotation like current vr, since I’m basically trying to get around the problem of loading tons of vr maps by having them all on the same map.

These are more meant to be like 100x100 maximum, preferably smaller sized minigames on one map. (like murderdome and syndicate trainer i guess)

**Comment by Kmc2000** [comment]
Sure, I'll do that instead and clean up the code now this is actually viable

**Comment by tralezab** [comment]
oh lordie this pr is going places and i only get to upvote it once

**Review by ShizCalev — DISMISSED** [review]

File: code/modules/minecraft/minecraft.dm [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	var/mob/steve
+	var/ores = list(/obj/item/stack/ore/iron,/obj/item/stack/ore/uranium,/obj/item/stack/ore/diamond,/obj/item/stack/ore/silver,/obj/item/stack/ore/plasma,/obj/item/stack/ore/bluespace_crystal,/obj/item/stack/ore/bananium)
+
+/turf/open/floor/attackby(obj/I, mob/user, params)
+	. = ..()
+	if(istype(I, /obj/item/minecraft))
+		to_chat(user, "You begin placing a block...")
+		if(do_after(user, 20, target = src))
+			var/obj/item/minecraft/X = I
+			to_chat(user, "you place [I]")
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: capitalize, period.

File: code/modules/minecraft/minecraft.dm [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	desc = "it mines things, very cool!"
+	icon = 'icons/obj/minecraft_items.dmi'
+	icon_state = "pickaxe"
+
+/turf/closed/wall/minecraft
+	name = "dirt block"
+	desc = "it's....totally square?"
+	icon = 'icons/turf/minecraft.dmi'
+	icon_state = "dirt"
+	var/strong = FALSE //can you mine this by hand? if false, you can
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: new vars at the bottom

File: code/modules/minecraft/minecraft.dm [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+/obj/item/minecraft/Initialize()
+	. = ..()
+	SpinAnimation(1000,1000)
+
+/turf/closed/wall/minecraft/attack_hand(mob/user)
+	if(strong)
+		to_chat(user, "<b>[src] is too strong to mine by hand!</b>")
+		return
+	if(isliving(user))
+		to_chat(user, "you start picking away at [src]")
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: capitalize

File: code/modules/minecraft/minecraft.dm [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	icon_state = "stonebrick"
+	build_type = /turf/closed/wall/minecraft/stonebrick
+
+/obj/item/minecraft/Initialize()
+	. = ..()
+	SpinAnimation(1000,1000)
+
+/turf/closed/wall/minecraft/attack_hand(mob/user)
+	if(strong)
+		to_chat(user, "<b>[src] is too strong to mine by hand!</b>")
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: span class

File: code/modules/minecraft/minecraft.dm [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		return
+	if(isliving(user))
+		to_chat(user, "you start picking away at [src]")
+		START_PROCESSING(SSfastprocess,src)
+		beingdug = TRUE
+		steve = user
+		if(do_after(user, 30, target = src))
+			var/sound = pick(hit_sounds)//Spam the mining sound minecraft style :)
+			playsound(loc,sound,50)
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: playsound(src

File: code/modules/minecraft/minecraft.dm [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+			new drop_type(src)
+			ChangeTurf(floor_type)
+		beingdug = FALSE
+		STOP_PROCESSING(SSfastprocess,src)
+		steve = null
+
+/turf/closed/wall/minecraft/attackby(obj/I, mob/user, params)
+	if(istype(I, /obj/item/pickaxe))
+		to_chat(user, "you start picking away at [src]")
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: capitalize

File: code/modules/minecraft/minecraft.dm [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	desc = "it's....totally square?"
+	icon_state = "stonebrick"
+	strong = TRUE
+	floor_type = /turf/open/floor/minecraft/stone
+	drop_type = /obj/item/minecraft/stonebrick
+	hit_sounds = list('sound/effects/minecraft/stone1.ogg','sound/effects/minecraft/stone2.ogg','sound/effects/minecraft/stone3.ogg','sound/effects/minecraft/stone4.ogg')
+
+/turf/closed/wall/minecraft/cobblestone
+	name = "cobblestone block"
+	desc = "it's....totally square?"
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: since this is the same for all of these, just have it on the parent obj

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Shadowflame909** [comment]
I think its pretty april fools day worthy

add it to the april fools day branch

**Comment by MASMC** [comment]
> I think its pretty april fools day worthy
> 
> add it to the april fools day branch

Why april fool's day, when it could be *every day?*

**Comment by Whoneedspacee** [comment]
i’m adding crappy minecraft minigames to start with for the vr hub anyways so this is fitting

**Comment by Kmc2000** [comment]
@Whoneedspacee 
Map's there if you wanna take the components for your own PR

File: code/modules/minecraft/minecraft.dm [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		steve = user
+		if(do_after(user, 20, target = src))
+			var/sound = pick(hit_sounds)//Spam the mining sound minecraft style :)
+			SEND_SOUND(user,sound)
+			if(strong)
+				if(prob(5))
+					var/mineralType = pick(ores)
+					var/num = rand(1,5)
+					new mineralType(src, num)
+			new drop_type(src)
```
Comment (SouDescolado) [inline code/modules/minecraft/minecraft.dm]: Either destroy the block on getting ores or make placed blocks not give ores.

File: code/modules/minecraft/minecraft.dm [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+					var/mineralType = pick(ores)
+					var/num = rand(1,5)
+					new mineralType(src, num)
+					return
+			new drop_type(src)
+			ChangeTurf(floor_type)
+		beingdug = FALSE
+		STOP_PROCESSING(SSfastprocess,src)
+		steve = null
+
```
Comment (SouDescolado) [inline code/modules/minecraft/minecraft.dm]: No, not like this. Do

```
if(strong && prob(5))
    drop ore code
else
    drop block code
changeturf
```

**Review by SouDescolado — CHANGES_REQUESTED** [review]

**Comment by obscolene** [comment]
the TGMC we need but not the one we deserve

**Comment by iskyp** [comment]
imagine going off to explore space and then coming back to galacticcraft

**Comment by Rowelll** [comment]
minecraft space when

**Comment by Zap-zapper** [comment]
GET OUT OF MY ROOM I'M PLAYING MINECRAFT

**Comment by Electronics** [comment]
I had to actually check it wasn't april

**Comment by ShizCalev** [comment]
```
[21:19:18] Runtime in minecraft.dm, line 68: undefined variable /turf/open/floor/minecraft/var/hit_sounds
 proc name: attack hand (/turf/closed/wall/minecraft/attack_hand)
usr: SylvinSeik/(Cynthia Elva)
usr.loc: (Lavaland Wastes (76, 42, 5))
src: the dirt block (77,43,5) (/turf/open/floor/minecraft)
call stack:
the dirt block (77,43,5) (/turf/open/floor/minecraft): attack hand(Cynthia Elva (/mob/living/carbon/human))
Cynthia Elva (/mob/living/carbon/human): UnarmedAttack(the dirt block (77,43,5) (/turf/open/floor/minecraft), 1)
Cynthia Elva (/mob/living/carbon/human): ClickOn(the dirt block (77,43,5) (/turf/open/floor/minecraft), "icon-x=26;icon-y=10;left=1;scr...")
the dirt block (77,43,5) (/turf/open/floor/minecraft): Click(the dirt block (77,43,5) (/turf/open/floor/minecraft), "mapwindow.map", "icon-x=26;icon-y=10;left=1;scr...")
```

**Comment by ShizCalev** [comment]
```
[21:14:50] Runtime in minecraft.dm, line 92: Undefined variable /turf/open/floor/minecraft/var/beingdug 
proc name: attackby (/turf/closed/wall/minecraft/attackby)
usr: SylvinSeik/(Cynthia Elva)
usr.loc: (Lavaland Wastes (73, 40, 5))
src: the dirt block (73,41,5) (/turf/open/floor/minecraft)
call stack:
the dirt block (73,41,5) (/turf/open/floor/minecraft): attackby(Pickaxe (/obj/item/pickaxe/minecraft), Cynthia Elva (/mob/living/carbon/human), "icon-x=24;icon-y=27;left=1;scr...")
Pickaxe (/obj/item/pickaxe/minecraft): melee attack chain(Cynthia Elva (/mob/living/carbon/human), the dirt block (73,41,5) (/turf/open/floor/minecraft), "icon-x=24;icon-y=27;left=1;scr...")
Cynthia Elva (/mob/living/carbon/human): ClickOn(the dirt block (73,41,5) (/turf/open/floor/minecraft), "icon-x=24;icon-y=27;left=1;scr...")
the dirt block (73,41,5) (/turf/open/floor/minecraft): Click(the dirt block (73,41,5) (/turf/open/floor/minecraft), "mapwindow.map", "icon-x=24;icon-y=27;left=1;scr...")
```

File: code/modules/minecraft/minecraft.dm [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+			var/sound = pick(hit_sounds)//Spam the mining sound minecraft style :)
+			SEND_SOUND(user,sound)
+			if(strong && prob(5))
+				var/mineralType = pick(ores)
+				var/num = rand(1,5)
+				new mineralType(src, num)
+			else
+				new drop_type(src)
+		ChangeTurf(floor_type)
+		beingdug = FALSE
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: the type has changed in the do_after before this, so this var is no longer a thing.

Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: the type has changed in the do_after before this, causing the runtime.

File: code/modules/minecraft/minecraft.dm [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+		to_chat(user, "You start picking away at [src].")
+		START_PROCESSING(SSfastprocess,src)
+		beingdug = TRUE
+		steve = user
+		if(do_after(user, 30, target = src))
+			var/sound = pick(hit_sounds)//Spam the mining sound minecraft style :)
+			playsound(src,sound,50)
+			new drop_type(src)
+			ChangeTurf(floor_type)
+		beingdug = FALSE
```
Comment (ShizCalev) [inline code/modules/minecraft/minecraft.dm]: .

**Comment by AlexTheSysop42** [comment]
Are the walls spaceproof?

File: code/modules/minecraft/minecraft.dm (hunk 10 above)
Comment (Kmc2000) [reply code/modules/minecraft/minecraft.dm]: Whoops!

**Comment by Bawhoppen** [comment]
Please add ruins (mob dungeons) and resprite tendrils to look like mob spawners

**Comment by ondrej008** [comment]
@Kmc2000 Make it so the blocks remember the floor that was under them when they were placed, so that when you destroy a block, the floor is the same as when you placed it. This way you can't transform the station into dirt.

**Comment by Shadowflame909** [comment]
Reicon the Ash Drake to the Elder Dragon and Bubblegum to Herobrine

**Comment by ondrej008** [comment]
@ExcessiveUseOfCobblestone Why did you close this without saying why you closed it?

**Comment by SouDescolado** [comment]
It was tagged as stale 4 days ago, with no comment from PR Author. If there is a development the PR Author may request a reopen.

**Comment by ExcessiveUseOfCobblestone** [comment]
because im hitler

**Comment by ondrej008** [comment]
> 
> 
> because im hitler

Hi Hitler, I'm dad.

**Pull request closed without merging**